src/shogun/clustering/KMeans.cpp

```diff
@@ -77,35 +77,31 @@ void CKMeans::set_initial_centers(SGMatrix<float64_t> centers)
 
 void CKMeans::set_random_centers(SGVector<float64_t> weights_set, SGVector<int32_t> ClList, int32_t XSize)
 {
+	/*initialise the cluster centroids randomly among all data*/
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
 
-	for (int32_t i=0; i<XSize; i++)
+	SGVector<int32_t> Perm=SGVector<int32_t>(XSize);
+	SGVector<int32_t>::range_fill_vector(Perm, XSize, 0);
+	CMath::permute(Perm);
+
+	for (int32_t i=0; i<k; i++)
 	{
-		const int32_t Cl=CMath::random(0, k-1);
+		const int32_t Cl=Perm[i];
 		weights_set[Cl]+=1.0;
-		ClList[i]=Cl;
+		ClList[Cl]=Cl;
 
 		int32_t vlen=0;
 		bool vfree=false;
-		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+		float64_t* vec=lhs->get_feature_vector(Cl, vlen, vfree);
```

Comment:
can you replace the get_feature_vector and free_feature_vector calls with the `SGVector` version. This is old code before we had `SGVector`, and the new version just works as ```SGVector<float64_t> vec = lhs->get_feature_vector(Cl)``` and doesnt require cleaning up afterwards